### PR TITLE
[NPM] Support pod controller to improve reliability

### PR DESCRIPTION
**Pull request by JungukCho**
**Reason for Change**:

This PR is to support pod controller in npm to improve reliability according to pod events.
For example, when event handling was failed, we need to retry them.

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests

**Notes**:
First put code layout to support pod controller. 
Will need to fill up business logics.

### CODE REVIEW

File: npm/podController.go [hunk 1]
```diff
@@ -165,8 +175,204 @@ func GetPodKey(podObj *corev1.Pod) string {
...
+				delete(c.podCache, key)
+			}
+			// for other transient apiserver error requeue with exponential backoff
+			return err
+		}
+	}
+
+	// TODO: do action for create and update events
+	// install ipset, update, and delete ipset
+	err = c.syncPod(pod)
```
Comment (neaggarwMS) [inline npm/podController.go:357]: This needs to be done within npmgr.Lock

File: npm/podController.go [hunk 2]
```diff
@@ -154,154 +163,353 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
+func (c *podController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
 		return nil
 	}
 
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
```
Comment (neaggarwMS) [inline npm/podController.go:377]: >podLister [](start = 15, length = 9)

Why do we need PodLister?

File: npm/podController.go [hunk 3]
```diff
@@ -154,154 +163,353 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
 		}
 		log.Logf("Creating set: %v, hashedSet: %v", newPodObjNs, util.GetHashedName(newPodObjNs))
 		if err = ipsMgr.CreateSet(newPodObjNs, append([]string{util.IpsetNetHashFlag})); err != nil {
 			metrics.SendErrorLogAndMetric(util.PodID, "[UpdatePod] Error creating ipset %s with err: %v", newPodObjNs, err)
+			return err
 		}
 	}
 
-	cachedPodObj, exists := npMgr.PodMap[podKey]
+	cachedPodObj, exists := c.npMgr.PodMap[podKey]
```
Comment (neaggarwMS) [inline npm/podController.go:508]: If you are handling the podKey in this PR, you also need to change internal ipset struct to cache IP per PODKey and not PodUID

File: npm/podController.go [hunk 4]
```diff
@@ -154,154 +163,353 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
-		npMgr.NsMap[podNs], err = newNs(podNs)
+	if _, exists := c.npMgr.NsMap[podNs]; !exists {
+		c.npMgr.NsMap[podNs], err = newNs(podNs)
 		if err != nil {
 			metrics.SendErrorLogAndMetric(util.PodID, "[AddPod] Error: failed to create namespace %s with err: %v", podNs, err)
 		}
 		log.Logf("Creating set: %v, hashedSet: %v", podNs, util.GetHashedName(podNs))
 		if err = ipsMgr.CreateSet(podNs, append([]string{util.IpsetNetHashFlag})); err != nil {
 			metrics.SendErrorLogAndMetric(util.PodID, "[AddPod] Error: creating ipset %s with err: %v", podNs, err)
+			return err
```
Comment (neaggarwMS) [inline npm/podController.go:442]: Can you skip this change, Vamsi PR is already taking care of it

File: npm/podController.go [hunk 5]
```diff
@@ -154,154 +163,353 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
 	log.Logf("Adding pod %s to ipset %s", podIP, podNs)
 	if err = ipsMgr.AddToSet(podNs, podIP, util.IpsetNetHashFlag, podUID); err != nil {
 		metrics.SendErrorLogAndMetric(util.PodID, "[AddPod] Error: failed to add pod to namespace ipset with err: %v", err)
+		return err
 	}
 
 	// Add the pod to its label's ipset.
 	for podLabelKey, podLabelVal := range podLabels {
 		log.Logf("Adding pod %s to ipset %s", podIP, podLabelKey)
 		if err = ipsMgr.AddToSet(podLabelKey, podIP, util.IpsetNetHashFlag, podUID); err != nil {
```
Comment (neaggarwMS) [inline npm/podController.go:456]: This should pass PodKey.

File: npm/podController.go (hunk 1 above)
Comment (JungukCho) [reply npm/podController.go]: Put Lock!

File: npm/podController.go (hunk 2 above)
Comment (JungukCho) [reply npm/podController.go:377]: We will get pods from shareInformer's local cache.

File: npm/podController.go (hunk 3 above)
Comment (JungukCho) [reply npm/podController.go:498]: Use podkey in ipset as well.

File: npm/podController.go [hunk 6]
```diff
@@ -146,162 +155,356 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
+		metrics.SendErrorLogAndMetric(util.PodID, err.Error())
+		return key, needSync
 	}
 
+	needSync = true
+	return key, needSync
+}
+
+func (c *podController) addPod(obj interface{}) {
+	log.Logf("[POD ADD EVENT]")
```
Comment (vakalapa) [inline npm/podController.go:229]: Move these event logs in needsycn

File: npm/podController.go (hunk 5 above)
Comment (JungukCho) [reply npm/podController.go]: Use podKey instead of podUID

File: npm/podController.go (hunk 6 above)
Comment (JungukCho) [reply npm/podController.go]: Update code with more contexts

File: npm/podController.go [hunk 7]
```diff
@@ -146,162 +155,356 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
+				if err != nil {
+					// cleaning process was failed, need to requeue and retry later.
+					return fmt.Errorf("Cannot delete ipset due to %s\n", err.Error())
+				}
+			}
+		}
+		return err
 	}
 
+	err = c.syncAddAndUpdatePod(pod)
```
Comment (vakalapa) [inline npm/podController.go:405]: before this we should check for deletetimestamp and clean up



	if pod.DeletionTimestamp != nil ||
		pod.DeletionGracePeriodSeconds != nil {
		err := c.cleanUpDeletedPod(pod)
return err

	}

File: npm/podController.go [hunk 8]
```diff
@@ -146,162 +155,356 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
...
-		return podErr
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	// lock to complete events
+	// TODO: Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("pod '%s' in work queue no longer exists", key))
```
Comment (vakalapa) [inline npm/podController.go:390]: Not an error, but an info message as Deleting Pod

File: npm/podController.go (hunk 7 above)
Comment (JungukCho) [reply npm/podController.go:425]: applied

File: npm/podController.go (hunk 8 above)
Comment (JungukCho) [reply npm/podController.go]: applied

**Comment by JungukCho** [comment]
Close this PR and create a new one ([NPM] Support pod controller and unit test to improve reliability #836
) to resovle conflict issue easily.

**Pull request closed without merging**